Periodic Expenses

bsonntag opened the pull request:
Adds Periodic Expenses, used to trigger periodic creation of Expenses.

Also adds a rake task (periodic_expenses:run) that goes through all periodic expenses. This task is scheduled to run every day.

bsonntag committed:
Adds PeriodicExpense model and tests

bsonntag committed:
Adds last_paid_at to PeriodicExpense.
Changes datetime fields of PeriodicExpense to date type.

bsonntag committed:
Adds timecop and whenever gems.

bsonntag committed:
Adds period_as_time and ready_to_pay? methods to PeriodicExpense and tests for these methods. Updates PeriodicExpense tests.

bsonntag committed:
Adds PeriodicExpenseRunner class and tests.

bsonntag committed:
Adds rake task for periodic expenses and scheduler.

naps62 commented on app/models/ability.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+
+  def registered_user
+    can :manage, Expense
```
given that this PR is about PeriodicExpenses (although this ability feature should have been in a different PR, actually), shouldn't there be a `:manage, PeriodicExpense` here as well?

bsonntag replied:
I don't know how this got here... It's PR https://github.com/groupbuddies/botler/pull/2, I must have accidentally added the file to this branch. :(

gabrielpoca replied:
@bsonntag do you know about [this](http://nuclearsquid.com/writings/git-add/)?

naps62 commented on app/models/periodic_expense.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def period_as_time
+    PERIODS[period]
+  end
```
this method is only used internally, so it could probably be private
Additionally, it's only used once, and it's pretty small, so why not just inline it?

bsonntag replied:
I used it outside the class in a draft at first.

Good point, I'll inline it.

naps62 commented on app/models/periodic_expense.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def ready_to_pay?
+    !next_pay.future?
```
I prefer to follow the convention of having all dates suffixed with `_at`, so it's more clear that they're dates. So this would become `next_pay_at`.
`next_pay` may sound like an instance of an actual payment

This is only a suggestion of course. I believe this isn't even in the guides

bsonntag replied:
It's also a little easier to read "next pay at future?"

Actually, it's in the guides:

> - Name date columns with _on suffixes.
> - Name datetime columns with _at suffixes.

Following this (although it says it's for columns), the suffix should be `_on`, and the same applies to `last_paid_at`.

What about `start_date` and `end_date`?

naps62 replied:
those could also change to something like started_on and ended_on, but since they already have `date` in the name, I don't see that as much of a problem. They're fine that way as well

naps62 commented on app/models/periodic_expense.rb:
```diff
@@ -0,0 +1,50 @@
+  def last_paid_at_is_between_start_and_end_dates
+    return if last_paid_at.blank? || between_start_and_end_dates?
+
+    errors.add(:end_date, 'cannot be before start date')
```
shouldn't this error msg be about the `last_paid_at` field? Looks like a copy&paste error

btw, unit tests for each of these custom validations are also recommended

bsonntag replied:
Yes, it was a copy&paste error, sorry.

They already have tests: https://github.com/groupbuddies/botler/blob/bs-create-periodic-expenses/spec/models/periodic_expense_spec.rb

naps62 replied:
:+1:

naps62 commented on app/models/periodic_expense_runner.rb:
```diff
@@ -0,0 +1,28 @@
+  private
+
+  def periodic_expenses
+    PeriodicExpense.where('start_date <= :today AND (end_date IS NULL OR end_date > :today)', today: today)
```
this logic would probably be better as a scope on the model itself. For example `PeriodicExpense.current`

naps62 commented on app/models/periodic_expense_runner.rb:
```diff
@@ -0,0 +1,28 @@
+    name = "#{periodic_expense.name} (#{periodic_expense.period.downcase})"
+    periodic_expense.last_paid_at = today
+    expense = Expense.new(name: name, user: periodic_expense.user, date: today.to_datetime, value: periodic_expense.amount)
+    expense.save
```
1. You can just do `Expense.create` instead of calling `save right after`
2. Are you updating the periodic_expense at all? You're setting the new last_paid_at, but I don't see the model being saved anywhere
3. Assuming the save is missing, then both saves should be wrapped around a transaction (http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html), so that it becomes an atomic operation

bsonntag replied:
Didn't knew about this. So, would it be something like this?

``` ruby
periodic_expense.transaction do
  periodic_expense.save!
  expense.save!
end
```

naps62 replied:
exactly

naps62 commented on spec/models/periodic_expense_runner_spec.rb:
```diff
@@ -0,0 +1,61 @@
+
+          Timecop.freeze(last_pay.next_week)
+
+          expect { runner.run }.to change { Expense.count }.by(0)
```
tip: this can also be written as:
`expect { runner.run }.not_to change { Expense.count }`

not too important, just a suggestion for readability

bsonntag replied:
I didn't like the `by(0)` either. I'll change it.

naps62 commented on spec/models/periodic_expense_runner_spec.rb:
```diff
@@ -0,0 +1,61 @@
+      end
+    end
+  end
+end
```
what about the following cases:
- periodic expenses with a start date in the future
- periodic expenses with an end date in the past (aka disabled, should not be paid anymore)

bsonntag replied:
Should I also add periodic expenses without a `last_paid_at` (aka first pay)?

naps62 replied:
yes, good catch

naps62 commented on spec/models/periodic_expense_spec.rb:
```diff
@@ -0,0 +1,126 @@
+      expect(periodic_expense).not_to be_valid
+    end
+
+    context 'end_date' do
```
do you mean `'#end_date'`?

bsonntag committed:
Adds validates_timeliness gem.

bsonntag committed:
Renames last_paid_at to last_paid_on.
Removes period_as_time method from PeriodicExpense.
Adds 'current' scope to PeriodicExpense.
Adds at_beginning_of_period method to PeriodicExpense.
Adds run_for_periodic_expense method to PeriodicExpenseRunner.
Changes PeriodicExpense validations to use validates_timeliness gem.

bsonntag committed:
Cleans PeriodicExpenseRunner tests.

zamith committed:
Refactors some of the tests to take advantage of mocks

bsonntag committed:
Renames PeriodicExpense#ready_to_pay? to PeriodicExpense#due?. Adds test for PeriodicExpense.current. Refactors PeriodicExpense#run tests.

bsonntag committed:
Refactors some PeriodicExpense and PeriodicExpenseRunner tests.

Adds doubles to PeriodicExpense#run test.

zamith commented:
:+1:
